include/ginkgo/core/preconditioner/ilu.hpp

```diff
@@ -0,0 +1,377 @@
+    {
+        dim<2> expected_size =
+            ReverseApply ? dim<2>{u_solver_->get_size()[0], b->get_size()[1]}
+                         : dim<2>{l_solver_->get_size()[0], b->get_size()[1]};
+        if (cache_.intermediate == nullptr ||
+            cache_.intermediate->get_size() != expected_size) {
+            cache_.intermediate = matrix::Dense<ValueType>::create(
+                this->get_executor(), expected_size);
+        }
+    }
+
+    /**
+     * @internal  Looks at the build() method to determine the type of the
+     *            factory.
+     */
+    template <typename T>
+    using factory_type_t = decltype(T::build());
+
+    // Parameter type of function `with_criteria`.
+    using with_criteria_param_type =
+        std::shared_ptr<const stop::CriterionFactory>;
+
+    /**
+     * Helper structure to test if the Factory of SolverType has a function
+     * `with_criteria`.
+     *
+     * Contains a constexpr boolean `value`, which is true if the Factory class
+     * of SolverType has a `with_criteria`, and false otherwise.
+     *
+     * @tparam SolverType   Solver to test if its factory has a with_criteria
+     *                      function.
+     *
+     */
+    template <typename SolverType, typename = void>
+    struct has_with_criteria : std::false_type {};
+
+    /**
+     * @copydoc has_with_criteria
+     *
+     * @internal  The second template parameter (which uses SFINAE) must match
+     *            the default value of the general case in order to be accepted
+     *            as a specialization.
+     */
+    template <typename SolverType>
+    struct has_with_criteria<
+        SolverType,
+        xstd::void_t<decltype(std::declval<factory_type_t<SolverType>>()
+                                  .with_criteria(with_criteria_param_type()))>>
+        : std::true_type {};
+
+
+    /**
+     * Generates a default solver of type SolverType.
+     *
+     * Also checks wheather SolverType can be assigned a criteria, and if it
+     * can, it is assigned default values which should be well suited for a
+     * preconditioner.
+     */
+    template <typename SolverType>
+    static xstd::enable_if_t<has_with_criteria<SolverType>::value,
+                             std::unique_ptr<SolverType>>
+    generate_default_solver(const std::shared_ptr<const Executor> &exec,
+                            const std::shared_ptr<const LinOp> &mtx)
+    {
+        constexpr ValueType default_reduce_precision{1e-4};
+        const unsigned int default_max_iters{
+            static_cast<unsigned int>(mtx->get_size()[0])};
+
+        return SolverType::build()
+            .with_criteria(gko::stop::Iteration::build()
+                               .with_max_iters(default_max_iters)
+                               .on(exec),
+                           gko::stop::ResidualNormReduction<>::build()
+                               .with_reduction_factor(default_reduce_precision)
```

Comment:
This should be `default_reduce_residual`. `default_reduce_precision` would mean something completely different.